src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,502 @@
+        }
+
+        // Set hash of rangePayment in mapping
+        isRangePaymentHash[msg.sender][rangePaymentHash] = true;
+        paymentNonce++;
+
+        rangePayment.token.safeTransferFrom(msg.sender, address(this), rangePayment.amount);
+        return rangePaymentHash;
+    }
+
+    /**
+     * @notice Process a payment claim.
+     * @param claim PaymentMerkleClaim struct containing the claim details for proof verification
+     * and payment transfer.
+     * @dev only callable by claimerFor[claim.earner]
+     */
+    function _processClaim(PaymentMerkleClaim calldata claim) internal onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) {
+        require(
+            _checkClaim(claim),
+            "PaymentCoordinator._processClaim: claim does not pass the check"
+        );
+
+        address claimer = claimerFor[claim.earnerLeaf.earner];
+        if (claimer == address(0)) {
+            claimer = claim.earnerLeaf.earner;
+        }
+        for (uint256 i = 0; i < claim.tokenIndices.length; i++) {
+            _processTokenClaim({
+                earnerLeaf: claim.earnerLeaf,
+                tokenLeaf: claim.tokenLeaves[i],
+                claimer: claimer,
+                root: distributionRoots[claim.rootIndex].root
+            });
+        }
+    }
+
+    /**
+     * @notice Transfer claimable amount to the claimer
+     * @param earnerLeaf leaf of earner merkle tree containing the earner address and earner's token root hash
+     * @param tokenLeaf token leaf to be claimed
+     * @param claimer address of the entity that can claim payments on behalf of the earner,
+     * can be earner account itself or be set to a different address by the earner
+     * @param root distribution root that should be read from storage
+     * @dev This function assumes the claim has already been checked and verified
+     */
+    function _processTokenClaim(
+        EarnerTreeMerkleLeaf calldata earnerLeaf,
+        TokenTreeMerkleLeaf calldata tokenLeaf,
+        address claimer,
+        bytes32 root
+    ) internal {
+        // Calculate amount to claim and update cumulativeClaimed
+        // Will revert if new leaf cumulativeEarnings is less than cumulative claimed
+        uint256 claimAmount = tokenLeaf.cumulativeEarnings - cumulativeClaimed[earnerLeaf.earner][tokenLeaf.token];
+        cumulativeClaimed[earnerLeaf.earner][tokenLeaf.token] = tokenLeaf.cumulativeEarnings;
+
+        tokenLeaf.token.safeTransfer(claimer, claimAmount);
+        emit PaymentClaimed(root, tokenLeaf);
+    }
+
+    function _checkClaim(PaymentMerkleClaim calldata claim) internal view returns (bool) {
+        DistributionRoot memory root = distributionRoots[claim.rootIndex];
+        require(block.timestamp >= root.activatedAt, "PaymentCoordinator._checkClaim: root not activated yet");
+
+        // If claimerFor earner is not set, claimer is by default the earner. Else set to claimerFor
+        address claimer = claimerFor[claim.earnerLeaf.earner];
```

Comment:
we do this twice in `_processClaim` also, let's only do it in `_processClaim`

Reply:
Also this msg.sender check means it would revert if calling the view `checkClaim` function. Even if the claim is valid and passes all the merkle proofs. Going to move this out of the _checkClaim internal function checks.

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/488/commits/a27911bceb4c60db7ad1a53551301dfb4615d020#diff-2a28246efa54c09dff5f0afef55f490d130b1a13c5b7e30efb51d5647c52a1a2L279
Now checkClaim will pass if proofs are valid. Assuming the caller is the valid claimer